### PR TITLE
cross compile binaries / dockerfile fixes

### LINKED ISSUE #74: cross compile notary cmd binaries
Should probably include a make target for cross compiling static binaries that can then be hosted somewhere.

you could even go as far as creating debs and rpms for hosting on our apt/yum repos ( but I need to work on adding those to runc first )

it should be easy for someone to download a binary for the notary command without installing go

Also I'm calling dibs on this but I opened this in case this is something you don't want lol, also you don't have a main Dockerfile anymore which I'm going to fix too because hell if I'm installing the cross compile crap on my host

### PR DESCRIPTION
**Pull request by jfrazelle**
closes #74 

1. Re-add back root Dockerfile for cross compiling and with GOPATH setup correctly.

2. Fix notary-server-dockerfile so the COPY is after the run otherwise the cache
will bust on the COPY and you wind up having to run the RUN apt-get every
single time you build the Dockerfile.

3. fix gross as eff RUN in notary-signer-dockerfile and wow really `apt-get update` on one line...

4. Add makefile target for cross. Can be run with:

```
docker build --rm --force-rm -t notary .

docker run --rm -it \
    -v $(pwd)/cross:/go/src/github.com/docker/notary/cross \
    notary \
    make cross
```

### CODE REVIEW

**Comment by jfrazelle** [comment]
```
$ docker run --rm -it     -v $(pwd)/cross:/go/src/github.com/docker/notary/cross     notary     make cross
mkdir -p /go/src/github.com/docker/notary/cross/darwin/amd64;
GOOS=darwin GOARCH=amd64 go build -o /go/src/github.com/docker/notary/cross/darwin/amd64/notary -a -tags "static_build netgo" -installsuffix netgo -ldflags "-w -extldflags -static -X `go list ./version`.Version `git describe --match 'v[0-9]*' --dirty='.m' --always`" ./cmd/notary; mkdir -p /go/src/github.com/docker/notary/cross/freebsd/amd64;
GOOS=freebsd GOARCH=amd64 go build -o /go/src/github.com/docker/notary/cross/freebsd/amd64/notary -a -tags "static_build netgo" -installsuffix netgo -ldflags "-w -extldflags -static -X `go list ./version`.Version `git describe --match 'v[0-9]*' --dirty='.m' --always`" ./cmd/notary; mkdir -p /go/src/github.com/docker/notary/cross/linux/amd64;
GOOS=linux GOARCH=amd64 go build -o /go/src/github.com/docker/notary/cross/linux/amd64/notary -a -tags "static_build netgo" -installsuffix netgo -ldflags "-w -extldflags -static -X `go list ./version`.Version `git describe --match 'v[0-9]*' --dirty='.m' --always`" ./cmd/notary; mkdir -p /go/src/github.com/docker/notary/cross/windows/amd64;
GOOS=windows GOARCH=amd64 go build -o /go/src/github.com/docker/notary/cross/windows/amd64/notary -a -tags "static_build netgo" -installsuffix netgo -ldflags "-w -extldflags -static -X `go list ./version`.Version `git describe --match 'v[0-9]*' --dirty='.m' --always`" ./cmd/notary; mkdir -p /go/src/github.com/docker/notary/cross/darwin/386;
GOOS=darwin GOARCH=386 go build -o /go/src/github.com/docker/notary/cross/darwin/386/notary -a -tags "static_build netgo" -installsuffix netgo -ldflags "-w -extldflags -static -X `go list ./version`.Version `git describe --match 'v[0-9]*' --dirty='.m' --always`" ./cmd/notary; mkdir -p /go/src/github.com/docker/notary/cross/freebsd/386;
GOOS=freebsd GOARCH=386 go build -o /go/src/github.com/docker/notary/cross/freebsd/386/notary -a -tags "static_build netgo" -installsuffix netgo -ldflags "-w -extldflags -static -X `go list ./version`.Version `git describe --match 'v[0-9]*' --dirty='.m' --always`" ./cmd/notary; mkdir -p /go/src/github.com/docker/notary/cross/linux/386;
GOOS=linux GOARCH=386 go build -o /go/src/github.com/docker/notary/cross/linux/386/notary -a -tags "static_build netgo" -installsuffix netgo -ldflags "-w -extldflags -static -X `go list ./version`.Version `git describe --match 'v[0-9]*' --dirty='.m' --always`" ./cmd/notary; mkdir -p /go/src/github.com/docker/notary/cross/windows/386;
GOOS=windows GOARCH=386 go build -o /go/src/github.com/docker/notary/cross/windows/386/notary -a -tags "static_build netgo" -installsuffix netgo -ldflags "-w -extldflags -static -X `go list ./version`.Version `git describe --match 'v[0-9]*' --dirty='.m' --always`" ./cmd/notary;

$ ldd cross/darwin/amd64/notary 
	not a dynamic executable

$ ldd cross/linux/amd64/notary 
	not a dynamic executable
```

**Comment by jfrazelle** [comment]
```
$ tree cross
cross
├── darwin
│   ├── 386
│   │   └── notary
│   └── amd64
│       └── notary
├── freebsd
│   ├── 386
│   │   └── notary
│   └── amd64
│       └── notary
├── linux
│   ├── 386
│   │   └── notary
│   └── amd64
│       └── notary
└── windows
    ├── 386
    │   └── notary
    └── amd64
        └── notary
```

File: notary-signer-Dockerfile [hunk 1]
```diff
@@ -8,9 +8,23 @@ ENV SOPIN="1234"
...
+	libtool \
+	gtk-doc-tools \
+	gengetopt \
+	help2man \
+	libpcsclite-dev \
+	libzip-dev \
+	opensc \
+	libssl-dev \
+	usbutils \
+	vim \
```
Comment (jfrazelle) [inline notary-signer-Dockerfile]: why do u need vim ...

Comment (diogomonica) [inline notary-signer-Dockerfile]: we don't feel free to remove it.

**Comment by diogomonica** [comment]
LGTM, thank you for the help.

File: notary-signer-Dockerfile [hunk 2]
```diff
@@ -8,9 +7,13 @@ ENV SOPIN="1234"
 ENV LIBDIR="/usr/local/lib/softhsm/"
 
 # Install openSC and dependencies
-RUN apt-get update
-
-RUN apt-get install -y build-essential autoconf automake libtool gtk-doc-tools gengetopt help2man libpcsclite-dev libzip-dev opensc libssl-dev usbutils vim
```
Comment (jfrazelle) [inline notary-signer-Dockerfile]: you might want to make sure this wasnt too aggressive

**Comment by diogomonica** [comment]
Rebuilt using docker-compose build with these new changes and docker-compose up. Everything seems to check out.

Thank you.

**Pull request merged**